pippy/ManualPipelineStage.py

```diff
@@ -0,0 +1,465 @@
+        ]
+
+        logger.debug(
+            f"""
+            Rank: {pg_rank}
+            Stage id: {stage_id}
+            Stage num stages: {stage.num_stages}
+            Stage rank: {rank}
+            Stage world size: {world_size}
+            Stage {virtual_id * world_size}-{(virtual_id + 1) * world_size - 1} input shapes: {stage_input_shapes}
+            Stage {virtual_id * world_size}-{(virtual_id + 1) * world_size - 1} output shapes: {stage_output_shapes}
+        """
+        )
+
+        all_inputs.extend(stage_input_shapes)
+        all_outputs.extend(stage_output_shapes)
+
+    # log only rank 0's view, they will all be equivalent
+    if pg_rank == 0:
+        logger.info(
+            f"""
+            all stage inputs: {all_inputs}
+            all stage outputs: {all_outputs}
+        """
+        )
+
+    # Check if the output for stage 0 matches the input at stage 1, and so forth
+    for i in range(virtual_pipeline_size * world_size - 1):
+        if (out := all_outputs[i]) != (inp := all_inputs[i + 1]):
+            raise ValueError(
+                f"Stage_id {stage_id} output shape {out} at does not match stage_id {i + 1} input shape {inp}."
+            )
+
+
+class ManualPipelineStage(PipelineStageBase):
+    def __init__(
+        self,
+        module: nn.Module,
+        stage_id: int,
+        num_stages: int,
+        device: torch.device,
+        input_args: Optional[Union[torch.Tensor, List[torch.tensor]]] = None,
+        output_args: Optional[Union[torch.Tensor, List[torch.tensor]]] = None,
+    ):
+        super().__init__(stage_id, num_stages, device)
+        self.module = module.to(device)
+        self.is_first_stage = stage_id == 0
+        self.is_last_stage = stage_id == num_stages - 1
+        # When we materialize the model partition on cuda, we call reset_parameters() if it is available
+        # logger.info(f"input args {input_args=}")
+        self.inputs: List[torch.tensor] = []
+        self.outputs: List[torch.tensor] = []
+
+        if input_args is not None:
+            self.inputs = create_buffers(input_args, device)
+
+        if output_args is None:
+            self.outputs = self.module(*self.inputs)
+            # create buffers for the output so that the data is in the correct
+            # shape in order to use in p2p op (send)
+            self.outputs = create_buffers(self.outputs, device)
+        else:
+            self.outputs = create_buffers(output_args, device)
+
+        # this is used in backward
+        self.inputs_outputs: Deque[
+            Tuple[List[torch.tensor], List[torch.tensor]]
+        ] = deque()
+
+        # these are the buffers used in backwards send/recv, they are allocated later
+        self.inputs_grad: List[torch.tensor] = []
+        self.outputs_grad: List[torch.tensor] = []
+
+        # TODO: Calculating stage index from rank is not ideal, e.g. won't match in Interleaved 1F1B case.
```

Comment:
nit: fix this TODO?

Reply:
will fix by passing in PG into pipeline stage